### PR TITLE
OpenJPEG: decoder for J2K codestreams

### LINKED ISSUE #19083: cv::imread/imdecode cannot load JPEG2000 codestream
##### System information (version)

- OpenCV => 4.4
- Operating System / Platform => MacOS Big Sur
- Compiler => Clang 12

##### Detailed description

The JPEG2000 standard defines both a "codestream" format and a file format for JP2K data. A codestream should ideally always be wrapped in the file format container, but in practice raw codestreams are sometimes found in the wild.

The OpenCV integration of OpenJPEG can read the JP2K file format, but not raw codestreams (also checked today's master). OpenJPEG itself has no problem reading codestreams (tested with the `opj_decompress` tool).

A codestream can be identified by its [4-byte header](http://fileformats.archiveteam.org/wiki/JPEG_2000_codestream#Identification) `FF 4F FF 51`.

OpenCV's JP2K file signature check should test for both the file header and the codestream header.

##### Steps to reproduce

Call `cv::imread` on the attached file.
[balloon.j2c.zip](https://github.com/opencv/opencv/files/5680574/balloon.j2c.zip)

### PR DESCRIPTION
**Pull request by tailsu**
**Merge with extra**: https://github.com/opencv/opencv_extra/pull/833

Fixes #19083

<cut/>

### Pull Request Readiness Checklist

See details at https://github.com/opencv/opencv/wiki/How_to_contribute#making-a-good-pull-request

- [x] I agree to contribute to the project under Apache 2 License.
- [x] To the best of my knowledge, the proposed patch is not based on a code under GPL or other license that is incompatible with OpenCV
- [x] The PR is proposed to proper branch
- [x] There is reference to original bug report and related work
- [x] There is accuracy test, performance test and test data in opencv_extra repository, if applicable
      Patch to opencv_extra has the same branch name.
- [x] The feature is well documented and sample code can be built with the project CMake

```
force_builders=Linux AVX2
```

### CODE REVIEW

File: modules/imgcodecs/test/test_grfmt.cpp [hunk 1]
```diff
@@ -78,7 +78,9 @@ const string all_images[] =
     "readwrite/Bretagne2.jp2",
     "readwrite/Grey.jp2",
     "readwrite/Grey.jp2",
+    "readwrite/balloon.j2c",
```
Comment (alalek) [inline modules/imgcodecs/test/test_grfmt.cpp:81]: GDAL has issue with this file.
Need to update `notForGDAL` below.

File: modules/imgcodecs/src/grfmt_jpeg2000_openjpeg.hpp [hunk 2]
```diff
@@ -79,6 +78,23 @@ class Jpeg2KOpjDecoder CV_FINAL : public BaseImageDecoder
...
+
+class Jpeg2KJP2OpjDecoder CV_FINAL : public Jpeg2KOpjDecoderBase {
+public:
+    Jpeg2KJP2OpjDecoder();
+     ~Jpeg2KJP2OpjDecoder() CV_OVERRIDE = default;
+
+    ImageDecoder newDecoder() const CV_OVERRIDE;
+};
+
+class Jpeg2KJ2KOpjDecoder CV_FINAL : public Jpeg2KOpjDecoderBase {
```
Comment (alalek) [inline modules/imgcodecs/src/grfmt_jpeg2000_openjpeg.hpp:92]: > J2K

j2c ?

**Review by alalek — COMMENTED** [review]
Thank you for contribution!

Please take a look on comments below.

File: modules/imgcodecs/src/grfmt_jpeg2000_openjpeg.hpp [hunk 3]
```diff
@@ -61,13 +61,12 @@ using ImagePtr = std::unique_ptr<opj_image_t, detail::OpjImageDeleter>;
 
 } // namespace detail
 
-class Jpeg2KOpjDecoder CV_FINAL : public BaseImageDecoder
+class Jpeg2KOpjDecoderBase : public BaseImageDecoder
```
Comment (VadimLevin) [inline modules/imgcodecs/src/grfmt_jpeg2000_openjpeg.hpp:64]: I think it is better to move this class to `detail` namespace - it is not supposed to be instantiated directly

File: modules/imgcodecs/src/grfmt_jpeg2000_openjpeg.hpp [hunk 4]
```diff
@@ -61,13 +61,12 @@ using ImagePtr = std::unique_ptr<opj_image_t, detail::OpjImageDeleter>;
...
 {
 public:
-    Jpeg2KOpjDecoder();
-     ~Jpeg2KOpjDecoder() CV_OVERRIDE = default;
+    Jpeg2KOpjDecoderBase(OPJ_CODEC_FORMAT format);
+     ~Jpeg2KOpjDecoderBase() CV_OVERRIDE = default;
```
Comment (VadimLevin) [inline modules/imgcodecs/src/grfmt_jpeg2000_openjpeg.hpp:68]: You can safely remove `default` override of destructor (same for derived classes).

File: modules/imgcodecs/src/grfmt_jpeg2000_openjpeg.cpp [hunk 5]
```diff
@@ -606,7 +600,8 @@ bool Jpeg2KOpjDecoder::readData( Mat& img )
     switch (image_->color_space)
     {
     case OPJ_CLRSPC_UNKNOWN:
```
Comment (VadimLevin) [inline modules/imgcodecs/src/grfmt_jpeg2000_openjpeg.cpp:602]: Add a `/* FALLTHRU */` comment for this case.

File: modules/imgcodecs/src/grfmt_jpeg2000_openjpeg.hpp (hunk 2 above)
Comment (tailsu) [reply modules/imgcodecs/src/grfmt_jpeg2000_openjpeg.hpp:92]: The OpenJPEG codec constant for codestreams is called `OPJ_CODEC_J2K` so I'm using `J2K` here for consistency.
Codestreams don't have an official extension and have several common extensions, like .j2k, .j2c and a few others.

File: modules/imgcodecs/src/grfmt_jpeg2000_openjpeg.hpp (hunk 3 above)
Comment (alalek) [reply modules/imgcodecs/src/grfmt_jpeg2000_openjpeg.hpp:64]: It is not a part of public headers, so it is not available externally.
So, `detail` namespace is not necessary here.

File: modules/imgcodecs/test/test_grfmt.cpp (hunk 1 above)
Comment (tailsu) [reply modules/imgcodecs/test/test_grfmt.cpp:81]: 👍

File: modules/imgcodecs/src/grfmt_jpeg2000_openjpeg.hpp (hunk 4 above)
Comment (tailsu) [reply modules/imgcodecs/src/grfmt_jpeg2000_openjpeg.hpp]: 👍

File: modules/imgcodecs/src/grfmt_jpeg2000_openjpeg.cpp (hunk 5 above)
Comment (tailsu) [reply modules/imgcodecs/src/grfmt_jpeg2000_openjpeg.cpp:604]: 👍

File: modules/imgcodecs/src/grfmt_jpeg2000_openjpeg.hpp (hunk 3 above)
Comment (tailsu) [reply modules/imgcodecs/src/grfmt_jpeg2000_openjpeg.hpp:62]: I've just moved it to detail. Should I roll back the change then?

Comment (alalek) [reply modules/imgcodecs/src/grfmt_jpeg2000_openjpeg.hpp:62]: As you wish. Both cases should be OK.

Comment (tailsu) [reply modules/imgcodecs/src/grfmt_jpeg2000_openjpeg.hpp:62]: Let's leave it like it is now.

**Review by VadimLevin — APPROVED** [review]

**Pull request merged**